mesonbuild/msubprojects.py

```diff
@@ -1,380 +1,414 @@
 
-def checkout(r, wrap, repo_dir, options):
-    if wrap.type != 'git' or not os.path.isdir(repo_dir):
-        return True
-    branch_name = options.branch_name if options.branch_name else wrap.get('revision')
-    if not branch_name:
-        # It could be a detached git submodule for example.
-        return True
-    mlog.log(f'Checkout {branch_name} in {wrap.name}...')
-    if git_checkout(repo_dir, branch_name, create=options.b):
-        git_show(repo_dir)
+    def update_svn(self):
+        if not os.path.isdir(self.repo_dir):
+            self.log('  -> Not used.')
+            return True
+        revno = self.wrap.get('revision')
+        p, out, _ = Popen_safe(['svn', 'info', '--show-item', 'revision', self.repo_dir])
+        current_revno = out
+        if current_revno == revno:
+            return True
+        if revno.lower() == 'head':
+            # Failure to do pull is not a fatal error,
+            # because otherwise you can't develop without
+            # a working net connection.
+            subprocess.call(['svn', 'update'], cwd=self.repo_dir)
+        else:
+            subprocess.check_call(['svn', 'update', '-r', revno], cwd=self.repo_dir)
         return True
-    return False
 
-def download(r, wrap, repo_dir, options):
-    mlog.log(f'Download {wrap.name}...')
-    if os.path.isdir(repo_dir):
-        mlog.log('  -> Already downloaded')
+    def update(self):
+        self.log(f'Updating {self.wrap.name}...')
+        if self.wrap.type == 'file':
+            return self.update_file()
+        elif self.wrap.type == 'git':
+            return self.update_git()
+        elif self.wrap.type == 'hg':
+            return self.update_hg()
+        elif self.wrap.type == 'svn':
+            return self.update_svn()
+        elif self.wrap.type is None:
+            self.log('  -> Cannot update subproject with no wrap file')
+        else:
+            self.log('  -> Cannot update', self.wrap.type, 'subproject')
         return True
-    try:
-        r.resolve(wrap.name, 'meson')
-        mlog.log('  -> done')
-    except WrapException as e:
-        mlog.log('  ->', mlog.red(str(e)))
+
+    def checkout(self):
+        if self.wrap.type != 'git' or not os.path.isdir(self.repo_dir):
+            return True
+        branch_name = self.options.branch_name if self.options.branch_name else self.wrap.get('revision')
+        if not branch_name:
+            # It could be a detached git submodule for example.
+            return True
+        self.log(f'Checkout {branch_name} in {self.wrap.name}...')
+        if self.git_checkout(branch_name, create=self.options.b):
+            self.git_show()
+            return True
         return False
-    return True
 
-def foreach(r, wrap, repo_dir, options):
-    mlog.log(f'Executing command in {repo_dir}')
-    if not os.path.isdir(repo_dir):
-        mlog.log('  -> Not downloaded yet')
+    def download(self):
+        self.log(f'Download {self.wrap.name}...')
```

Comment:
Some of these can take a long time and they will not be printed until the end of the run. So one possibility could be to change all these log messages to start with `f'{self.wrap.name}: '`. Then there's no need to introduce the log queue, only the lock is needed.

If desired, you could have all messages after the first start with `'  -> '` instead of `f'{self.wrap.name}: '` if execution is not parallel.

Reply:
The output would still be a mess if logs are all interleaved, even when prefixed with the subproject name, IMHO.

I added a commit that shows a progress line: `Progress: 56 / 69 (win-nasm, glib-networking, vpx, win-flex-bison-binaries, libepoxy, tinyalsa, ca...)`

Reply:
This is similar to running parellel tests.